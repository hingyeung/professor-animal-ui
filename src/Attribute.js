import React, { Component } from 'react';

class Attribute extends Component {
  constructor(props) {
    super(props);

    this.name = props.name;
  }

  render() {
    return (
      <div class="attribute-row">
        <div class="attribute-cell">
          {this.name}
        </div>
        <div class="attribute-cell">
          <label htmlFor={this.name + '-yes'}>YES</label>
          <input type="radio" name={this.name} value="yes" id={this.name + '-yes'}/>
        </div>
        <div class="attribute-cell">
          <label htmlFor={this.name + '-no'}>NO</label>
          <input type="radio" name={this.name} value="no" id={this.name + '-no'}/>
        </div>
      </div>
    )
  }
}

export default Attribute;