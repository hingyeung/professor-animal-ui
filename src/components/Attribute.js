import React, {Component} from 'react';

class Attribute extends Component {
  constructor(props) {
    super(props);

    this.props = props;
    this.attribute = props.attribute;
    this.type = props.type;
    this.onAttributeChange = props.onAttributeChange;
  }

  onChange(attrType, attrName, e) {
    this.onAttributeChange(attrType, attrName, e.target.value);
  }

  render() {
    const yesId = this.type + '-' + this.attribute.name + '-yes',
          noId = this.type + '-' + this.attribute.name + '-no';

    return (
      <tr className="attribute-row row">
        <td className="attribute-cell col-sm-6">
          { this.attribute.name }
        </td>
        <td className="attribute-cell col-sm-3">
          <input className="form-check-label" type="radio" name={ this.type + '-' + this.attribute.name } value="yes"
                 id={ yesId }
                 onChange={ (e) => this.onChange(this.type, this.attribute.name, e) }
                 defaultChecked={this.attribute.value}/>
          <label className="form-check-label" htmlFor={ yesId }>YES</label>
        </td>
        <td className="attribute-cell col-sm-3">
          <input className="form-check-label" type="radio" name={ this.type + '-' + this.attribute.name } value="no"
                 id={ noId } onChange={ (e) => this.onChange(this.type, this.attribute.name, e) }
                 defaultChecked={!this.attribute.value}/>
          <label className="form-check-label" htmlFor={ noId }>NO</label>
        </td>
      </tr>
    )
  }
}

export default Attribute;