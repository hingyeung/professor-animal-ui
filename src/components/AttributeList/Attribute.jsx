import React from 'react';

function Attribute(props) {
  const attribute = props.attribute;
  const type = props.type;
  const onAttributeChange = props.onAttributeChange;
  const yesId = props.type + '-' + props.attribute.name + '-yes',
    noId = props.type + '-' + props.attribute.name + '-no',
    // Initially this.state.attributeValue is undefined (before animals are loaded),
    // and it makes the radio button "uncontrolled", then later when this.state.attributeValue
    // is populated with either true or false and the radio button becomes "controlled".
    // ReactJS doesn't like it and would give warning. Making sure its value is either true
    // or false prevents the problem.
    // https://github.com/facebook/react/issues/6779#issuecomment-222162404
    yesIsChecked = props.attribute.value ? props.attribute.value : false;

  function onChange(e) {
    onAttributeChange(type, attribute.name, e.target.value);
  }

  return (
    <tr className="attribute-row row">
      <td className="attribute-cell col-sm-6">
        { attribute.name }
      </td>
      <td className="attribute-cell col-sm-3">
        <input className="form-check-label" type="radio" name={ type + '-' + attribute.name } value="yes"
               id={ yesId }
               onChange={ onChange }
               checked={ yesIsChecked }/>
        <label className="form-check-label" htmlFor={ yesId }>YES</label>
      </td>
      <td className="attribute-cell col-sm-3">
        <input className="form-check-label" type="radio" name={ type + '-' + attribute.name } value="no"
               id={ noId } onChange={ onChange }
               checked={ !yesIsChecked }/>
        <label className="form-check-label" htmlFor={ noId }>NO</label>
      </td>
    </tr>
  )
}

export default Attribute;