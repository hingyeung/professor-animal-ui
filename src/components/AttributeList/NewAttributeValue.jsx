import React from 'react';
import Attribute from 'models/Attribute';
import PropTypes from 'prop-types';

function NewAttributeValue(props) {

  const onNewAttributeAdded = props.onNewAttributeAdded;
  const attributeType = props.attributeType;
  const newAttributeIndex = props.newAttributeIndex;
  let input;

  function _addNewAttribute(attributeName) {
    if (!attributeName) {
      // TODO make the field red
      console.log('Empty attribute name');
      return;
    }

    onNewAttributeAdded(new Attribute(attributeName, true));
    input.value = '';
  }

  function _onAddClicked(e) {
    e.preventDefault();
    _addNewAttribute(input.value);
  }

  function _handleKeyPress(e) {
    if (e.key === 'Enter') {
      e.preventDefault();
      _addNewAttribute(input.value);
    }
  }

  return (
    <tr className="attribute-row row">
      <td className="attribute-cell col-sm-6">
        <input type="text" name={ attributeType + '-' + newAttributeIndex }
               className="form-control"
               ref={ ref => input = ref }
               onKeyPress={ _handleKeyPress }
          // onChange={ (e) => this.setState({newAttributeName: e.target.value}) }
        />
      </td>
      <td className="attribute-cell col-sm-6">
        <button className="btn btn-primary" onClick={ _onAddClicked }>Add</button>
      </td>
    </tr>
  )
}

NewAttributeValue.propTypes = {
  onNewAttributeAdded: PropTypes.func.isRequired,
  attributeType: PropTypes.string.isRequired,
  newAttributeIndex: PropTypes.number.isRequired
};

export default NewAttributeValue;