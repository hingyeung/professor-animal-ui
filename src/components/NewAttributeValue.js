import React, {Component} from 'react';
import Attribute from '../models/Attribute';

class NewAttributeValue extends Component {
  constructor(props) {
    super(props);

    this.onAdd = props.onAdd;
    this.attributeType = props.type;
    this.newAttributeIndex = props.newAttributeIndex;
    this.onAddClicked = this.onAddClicked.bind(this);
    this.state = {
      newAttributeName: ''
    };
  }

  onAddClicked(attributeType, attributeName) {
    const attribute = new Attribute(attributeName, true);
    this.onAdd(attributeType, attribute);
  }

  render() {
    return (
      <tr className="attribute-row row">
        <td className="attribute-cell col-sm-6">
          <input type="text" name={ this.attributeType + '-' + this.newAttributeIndex }
                 value={ this.state.newAttributeName } className="form-control"
                 onChange={ (e) => this.setState({newAttributeName: e.target.value}) }/>
        </td>
        <td className="attribute-cell col-sm-6">
          <button className="btn btn-primary" onClick={() => this.onAddClicked(this.attributeType, this.state.newAttributeName) }>Add</button>
        </td>
      </tr>
    )
  }
}

export default NewAttributeValue;