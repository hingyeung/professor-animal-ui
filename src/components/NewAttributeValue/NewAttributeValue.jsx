import React from 'react';
import Attribute from 'models/Attribute';
import PropTypes from 'prop-types';

function NewAttributeValue(props) {

  const onNewAttributeAdded = props.onNewAttributeAdded;
  const attributeType = props.attributeType;
  const newAttributeIndex = props.newAttributeIndex;
  const INVALID_VALUE_CSS_CLASS = 'is-invalid';
  let input;

  function _addNewAttribute(attributeName) {
    if (!attributeName) {
      input.classList.add(INVALID_VALUE_CSS_CLASS);
      return;
    }

    onNewAttributeAdded(new Attribute(attributeName, true));
    input.classList.remove(INVALID_VALUE_CSS_CLASS);
    input.value = '';
  }

  function _onAddClicked(e) {
    e.preventDefault();
    _addNewAttribute(input.value);
  }

  function _handleKeyPress(e) {
    if (e.key === 'Enter') {
      e.preventDefault();
      _addNewAttribute(input.value);
    }
  }

  return (
    <tr className={"attribute-row row attribute-group attribute-group-" + attributeType}>
      <td className="attribute-cell col-sm-6">
        <input type="text" name={ attributeType + '-' + newAttributeIndex }
               className="form-control new-attribute-value-input"
               ref={ ref => input = ref }
               onKeyPress={ _handleKeyPress }
        />
        <div className="invalid-feedback">
          Please provide an <span className="all-caps">{attributeType}</span> attribute.
        </div>
      </td>
      <td className="attribute-cell col-sm-6">
        <button className="btn btn-primary new-attribute-add-btn" onClick={ _onAddClicked }>Add</button>
      </td>
    </tr>
  )
}

NewAttributeValue.propTypes = {
  onNewAttributeAdded: PropTypes.func.isRequired,
  attributeType: PropTypes.string.isRequired,
  newAttributeIndex: PropTypes.number.isRequired
};

export default NewAttributeValue;