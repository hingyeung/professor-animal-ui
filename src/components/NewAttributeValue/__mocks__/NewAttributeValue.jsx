import React from 'react';
import Attribute from "models/Attribute";

const NewAttributeValue = (props) => {
  const addNewAttribute = (e) => {
    props.onNewAttributeAdded(new Attribute('new attribute', true));
  };

  return <button id="mock-new-attribute-value" onClick={addNewAttribute}>Add new attribute</button>
};

export default NewAttributeValue;