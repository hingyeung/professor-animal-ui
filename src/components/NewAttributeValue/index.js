import NewAttributeValue from './NewAttributeValue';

export default NewAttributeValue;