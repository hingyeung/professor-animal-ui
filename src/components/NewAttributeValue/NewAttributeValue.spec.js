import React from 'react';
import renderer from 'react-test-renderer';
import NewAttributeValue from "./NewAttributeValue";
import {mount} from "enzyme/build/index";
import Attribute from "../../models/Attribute";
import toJson from 'enzyme-to-json';

const MOCK_ON_NEW_ATTRIBUTE_ADDED = jest.fn();
const ATTRIBUTE_TYPE = 'attribute_type';
const NEW_ATTRIBUTE_INDEX = 1;

describe('NewAttributeValue', function () {
  beforeEach(function() {
    MOCK_ON_NEW_ATTRIBUTE_ADDED.mockClear();
  });

  it('should render properly', function () {
    const tree = renderer.create(<NewAttributeValue onNewAttributeAdded={ MOCK_ON_NEW_ATTRIBUTE_ADDED }
                                                    attributeType={ ATTRIBUTE_TYPE }
                                                    newAttributeIndex={ NEW_ATTRIBUTE_INDEX }/>);

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should call onNewAttributeAdded in props when the input field is populated and the Add button is clicked', function () {
    // mount, not shallow, must be used when testing component reference
    const wrapper = mount(<NewAttributeValue onNewAttributeAdded={ MOCK_ON_NEW_ATTRIBUTE_ADDED }
                                             attributeType={ ATTRIBUTE_TYPE }
                                             newAttributeIndex={ NEW_ATTRIBUTE_INDEX }/>);

    const newAttributeInput = wrapper.find('.new-attribute-value-input');
    // https://stackoverflow.com/a/46611142
    newAttributeInput.instance().value = "some new attribute";
    const addButton = wrapper.find('.new-attribute-add-btn');
    const clickEvent = {preventDefault: jest.fn()};
    addButton.simulate('click', clickEvent);

    expect(MOCK_ON_NEW_ATTRIBUTE_ADDED).toBeCalledWith(new Attribute('some new attribute', true));
    // wrapper.unmount();
  });

  it('should not call onNewAttributeAdded in props when the input field value is blank and Add button is clicked', function () {
    const wrapper = mount(<NewAttributeValue onNewAttributeAdded={ MOCK_ON_NEW_ATTRIBUTE_ADDED }
                                             attributeType={ ATTRIBUTE_TYPE }
                                             newAttributeIndex={ NEW_ATTRIBUTE_INDEX }/>);

    const addButton = wrapper.find('.new-attribute-add-btn');
    const clickEvent = {preventDefault: jest.fn()};
    addButton.simulate('click', clickEvent);

    expect(MOCK_ON_NEW_ATTRIBUTE_ADDED).not.toHaveBeenCalled();
    expect(toJson(wrapper)).toMatchSnapshot();
    // wrapper.unmount();
  });
});