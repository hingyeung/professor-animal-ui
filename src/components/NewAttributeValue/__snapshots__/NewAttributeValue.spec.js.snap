// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewAttributeValue should render properly 1`] = `
<tr
  className="attribute-row row"
>
  <td
    className="attribute-cell col-sm-6"
  >
    <input
      className="form-control new-attribute-value-input"
      name="attribute_type-1"
      onKeyPress={[Function]}
      type="text"
    />
  </td>
  <td
    className="attribute-cell col-sm-6"
  >
    <button
      className="btn btn-primary new-attribute-add-btn"
      onClick={[Function]}
    >
      Add
    </button>
  </td>
</tr>
`;
